File: hwemulator/src/main/cpp/hwemulatorbase/entity/HwGamePadEvent.cpp
//
// Created by mingyi.li on 2019/5/17.
//

#include "../include/HwGamePadEvent.h"
#include "Logcat.h"

HwGamePadEvent *HwGamePadEvent::create(int key, int action) {
    return new HwGamePadEvent(key, action);
}

HwGamePadEvent::HwGamePadEvent(int key, int action) : Object() {
    this->key = key;
    this->action = action;
//    toString();
}

HwGamePadEvent::~HwGamePadEvent() {

}

int HwGamePadEvent::getKey() {
    return key;
}

int HwGamePadEvent::getAction() {
    return action;
}

void HwGamePadEvent::toString() {
    string key;
    string action;
    switch (this->key) {
        case KEY_UP:
            key = "UP";
            break;
        case KEY_DOWN:
            key = "DOWN";
            break;
        case KEY_LEFT:
            key = "LEFT";
            break;
        case KEY_RIGHT:
            key = "RIGHT";
            break;
        case KEY_SELECT:
            key = "SELECT";
            break;
        case KEY_START:
            key = "START";
            break;
        case KEY_A:
            key = "A";
            break;
        case KEY_B:
            key = "B";
            break;
        case KEY_NONE:
        default:
            key = "NONE";
    }
    switch (this->action) {
        case ACTION_UP:
            action = "UP";
            break;
        case ACTION_DOWN:
            action = "DOWN";
            break;
        case ACTION_NONE:
        default:
            action = "NONE";
    }
    Logcat::i("HWEMULATOR", "key=%s, action=%s", key.c_str(), action.c_str());
}

File: hwemulator/src/main/java/com/lmy/emulator/entity/HwGamePadEvent.kt
package com.lmy.emulator.entity

class HwGamePadEvent private constructor(key: Int, action: Int) {
    private var handler: Long = 0

    init {
        handler = create(key, action)
    }

    fun getKey(): Int {
        if (handler > 0) {
            return getKey(handler)
        }
        return KEY_NONE
    }

    fun getAction(): Int {
        if (handler > 0) {
            return getAction(handler)
        }
        return ACTION_NONE
    }

    private external fun create(key: Int, action: Int): Long
    private external fun getKey(handle: Long): Int
    private external fun getAction(handle: Long): Int

    companion object {
        fun create(key: Int, action: Int): HwGamePadEvent {
            return HwGamePadEvent(key, action)
        }

        const val KEY_NONE = 0x000
        const val KEY_RIGHT = 0x001
        const val KEY_LEFT = 0x002
        const val KEY_UP = 0x003
        const val KEY_DOWN = 0x004
        const val KEY_SELECT = 0x005
        const val KEY_START = 0x006
        const val KEY_A = 0x007
        const val KEY_B = 0x008
        const val KEY_ESC = 0x009
        const val ACTION_NONE = 0x000
        const val ACTION_DOWN = 0x101
        const val ACTION_UP = 0x102
    }
}